show symbolic links as decorations in explorer

isidorn opened the pull request:
fixes #11539

I have adopted your idea in the `fileService` to first do an `lstat` and if only if it is a symbolic link to do an additional `stat` call.

As for visualisation to not do any icon crazyness I have decided to use decorations similar to what we do for errors and git decorations.
I went through all the ansi arrows and I personally like this one the most ⇲
Here's a list of unicode arrows https://unicode-table.com/en/sets/arrows-symbols/
After we get feedback from users we can decide to do some additional decorating like italic, but not for now imho and also the decoration api does not support italic

fyi @chryw
A nice icon overlay you provided looks great, we can look into adding that to the icon as step 2, after we merge this is and get some feedback.

![screen shot 2018-02-16 at 11 02 24](https://user-images.githubusercontent.com/1926584/36302826-32b12e64-130a-11e8-8dfc-23382adad91f.png)

bpasero commented on src/vs/platform/files/common/files.ts:
```diff
@@ -410,6 +410,11 @@ export interface IFileStat extends IBaseStat {
 	isDirectory: boolean;
 
+	/**
+	 * The resource is a simbolic link.
```
simbolic => symbolic

bpasero commented on src/vs/workbench/services/files/test/node/resolver.test.ts:
```diff
@@ -19,7 +19,7 @@ function create(relativePath: string): StatResolver {
 	let fsStat = fs.statSync(absolutePath);
 
-	return new StatResolver(uri.file(absolutePath), fsStat.isDirectory(), fsStat.mtime.getTime(), fsStat.size, void 0);
+	return new StatResolver(uri.file(absolutePath), fsStat.isSymbolicLink(), fsStat.isDirectory(), fsStat.mtime.getTime(), fsStat.size, void 0);
```
@isidorn can we have a test for this too? it looks like node.js allows to create symlinks (`fs.symlink`).

bpasero commented on src/vs/workbench/parts/files/electron-browser/views/explorerDecorationsProvider.ts:
```diff
@@ -31,14 +31,20 @@ export class ExplorerDecorationsProvider implements IDecorationsProvider {
 		}
+		if (fileStat && fileStat.isSymbolicLink) {
+			return {
+				tooltip: localize('symbolicLlink', "Symbolic link"),
```
Maybe better capital casing: `Symbolic Link`

bpasero commented on src/vs/workbench/parts/files/electron-browser/views/explorerDecorationsProvider.ts:
```diff
@@ -31,14 +31,20 @@ export class ExplorerDecorationsProvider implements IDecorationsProvider {
+		if (fileStat && fileStat.isSymbolicLink) {
+			return {
+				tooltip: localize('symbolicLlink', "Symbolic link"),
+				letter: '\u21f2'
```
Maybe there is a better symbol? Something that looks more like this:

![image](https://user-images.githubusercontent.com/900690/36370462-37c9cbac-155f-11e8-9737-cee867ba83c5.png)

bpasero commented on src/vs/workbench/services/files/node/fileService.ts:
```diff
@@ -1254,6 +1259,7 @@ export class StatResolver {
 						const childStat: IFileStat = {
 							resource: fileResource,
 							isDirectory: fileStat.isDirectory(),
+							isSymbolicLink: isSymbolicLink,
```
Can be simplified to: `isSymbolicLink`

bpasero commented on src/vs/workbench/services/files/node/fileService.ts:
```diff
@@ -913,7 +913,7 @@ export class FileService implements IFileService {
 
 		return pfs.stat(absolutePath).then(stat => {
-			return new StatResolver(resource, stat.isDirectory(), stat.mtime.getTime(), stat.size, this.options.verboseLogging ? this.options.errorLogger : void 0);
+			return new StatResolver(resource, false, stat.isDirectory(), stat.mtime.getTime(), stat.size, this.options.verboseLogging ? this.options.errorLogger : void 0);
```
This is a bit weird, why not introduce a new method `pfs.statLink` that is implemented like this:
* executes `lstat`
* executes `stat` only if the stat is a symbolic link
* returns something like `{ stat: stat, isSymbolicLink: boolean }`

This method could then also be used from within the resolver

bpasero commented on src/vs/workbench/services/files/node/fileService.ts:
```diff
@@ -1235,7 +1233,14 @@ export class StatResolver {
 
 					function stat(this: any): void {
-						fs.stat(fileResource.fsPath, this);
+						fs.lstat(fileResource.fsPath, (error, stat) => {
```
Use the new method from extfs/pfs for this.

isidorn commented:
@bpasero I have addressed all your comments. Please review when you get a chance
I have also found a sexy unicode character
Have to check it on linux and win though.

![screen shot 2018-02-20 at 12 23 00](https://user-images.githubusercontent.com/1926584/36421489-e6d89176-1638-11e8-8dd6-6376d60350cf.png)

bpasero commented on src/vs/workbench/services/files/node/fileService.ts:
```diff
@@ -912,8 +912,8 @@ export class FileService implements IFileService {
 
-		return pfs.stat(absolutePath).then(stat => {
-			return new StatResolver(resource, stat.isDirectory(), stat.mtime.getTime(), stat.size, this.options.verboseLogging ? this.options.errorLogger : void 0);
+		return pfs.statLink(absolutePath).then(statAndIsLink => {
```
Can be written nicer:

```typescript
return pfs.statLink(absolutePath).then(({ isSymbolicLink, stat }) => {
	return new StatResolver(resource, isSymbolicLink, stat.isDirectory(), stat.mtime.getTime(), stat.size, this.options.verboseLogging ? this.options.errorLogger : void 0);
});
```

isidorn replied:
Done

bpasero commented on src/vs/workbench/services/files/node/fileService.ts:
```diff
@@ -1235,7 +1233,10 @@ export class StatResolver {
 
 					function stat(this: any): void {
-						fs.stat(fileResource.fsPath, this);
+						extfs.statLink(fileResource.fsPath, (error: Error, statAndIsLink) => {
```
Can be nicer:

```typescript
extfs.statLink(fileResource.fsPath, (error: Error, { stat, isSymbolicLink }) => {
	isSymbolicLink = isSymbolicLink;
	this(null, stat);
});
```

isidorn replied:
Did not do this due to name clash of the isSymoblicLink variable a couple of lines up. Decided to leave like it is.

bpasero commented on src/vs/base/node/pfs.ts:
```diff
@@ -54,6 +54,16 @@ export function stat(path: string): TPromise<fs.Stats> {
 	return nfcall(fs.stat, path);
 }
 
+export function statLink(path: string): TPromise<{ stat: fs.Stats, isSymbolicLink: boolean }> {
```
@isidorn why is this method not simply using extfs? E.g. it should be as simple as:

```typescript
export function statLink(path: string): TPromise<string> {
	return nfcall(extfs. statLink, path);
}
```

isidorn replied:
@bpasero you are correct. I was not aware I can depend on `extfs` in `pfs`, but now I see it is already done
